Sync/BatchingClient.swift

```diff
@@ -0,0 +1,263 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import Foundation
+import Alamofire
+import Shared
+import XCGLogger
+import Deferred
+
+public class BatchingNotSupported: MaybeErrorType {
+    public let response: POSTResult
+
+    public var description: String {
+        return "Sync server does not support batching."
+    }
+
+    public init(response: POSTResult) {
+        self.response = response
+    }
+}
+
+private let log = Logger.syncLogger
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private let config: InfoConfiguration
+
+    private var records: [Record<T>] = []
```

Comment:
You already work with strings as soon as you get to `batchesFromRecords`, and that method immediately serializes. So instead of having this be `[Record<T>]`, I suggest having this be `[(GUID, String)]`; `addRecords` will serialize.